PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/serviceexplorer/azure/appservice/ProfileFlightRecordAction.java

```diff
@@ -0,0 +1,202 @@
+                    notifyUserWithErrorMessage("Invalid profile duration",
+                            "Cannot profile on app service due to "
+                                    + "configuration error: invalid "
+                                    + "duration");
+                    finishLatch.countDown();
+                    return;
+                } else {
+                    new Thread(() -> {
+                        doProfileFlightRecorder(progressIndicator,
+                                                config, finishLatch);
+                    }).start();
+                }
+            }, ModalityState.NON_MODAL);
+            finishLatch.await();
+        } catch (Exception ex) {
+            notifyUserWithErrorMessage(
+                    "Failed to profile on app service",
+                    "Cannot profile on app service, due to error: "
+                            + ex.getMessage());
+        }
+    }
+
+    private FlightRecorderConfiguration collectFlightRecorderConfiguration() {
+        RunFlightRecorderDialog ui = new RunFlightRecorderDialog(
+                project,
+                appService);
+        ui.setTitle(String.format("Start Flight Recorder on '%s'", appService.defaultHostName()));
+        ui.setOkActionListener((config) -> {
+            ui.close(DialogWrapper.OK_EXIT_CODE);
+        });
+
+        if (ui.showAndGet()) {
+            return ui.getData();
+        }
+        return null;
+    }
+
+    private void doProfileFlightRecorder(ProgressIndicator progressIndicator,
+                                         FlightRecorderConfiguration config, CountDownLatch latch) {
+        try {
+            progressIndicator.setText(String.format("start flight recorder for process(%d:%s)",
+                    config.getPid(),
+                    config.getProcessName()));
+            File file = File.createTempFile("jfr-snapshot-" + appService.name() + "-", ".jfr");
+            FlightRecorderStarterBase starter = FlightRecorderManager.getFlightRecorderStarter(appService);
+            starter.startFlightRecorder(config.getPid(), config.getDuration(), file.getName());
+            progressIndicator.setText(String.format("Recording %s seconds...", config.getDuration()));
+
+            progressIndicator.checkCanceled();
+            try {
+                Thread.sleep(ONE_SECOND * config.getDuration() + TWO_SECONDS);
+            } catch (InterruptedException e) {
+                // ignore
+            }
+            progressIndicator.checkCanceled();
+            progressIndicator.setText("Profile completed on server side.");
+            progressIndicator.setText("Downloading jfr file...");
+            byte[] content = starter.downloadJFRFile(file.getName());
+            if (content != null) {
+                FileUtils.forceDeleteOnExit(file);
+                FileUtils.writeByteArrayToFile(file, content);
+                progressIndicator.setText("Download jfr file complete");
```

Comment:
do we need to show the complete here? as the progress will terminated soon

Reply:
no need but this message itself has no bad points